src/gui/widgets/settings.cpp

```diff
@@ -51,13 +51,16 @@ std::vector<game_tip> tips;
 
 void update_screen_size_variables()
 {
-	CVideo& vid = CVideo::get_singleton();
-	const SDL_Rect rect = vid.screen_area();
+	CVideo& video = CVideo::get_singleton();
+	const SDL_Rect rect = video.draw_area();
 
 	screen_width = rect.w;
 	screen_height = rect.h;
 
-	auto [scalew, scaleh] = vid.get_dpi_scale_factor();
+	// This should, perhaps, include the pixel scale multiplier.
+	// However i'm not sure what it is intended to be used for,
+	// and it doesn't appear to be currently used anywhere.
+	auto [scalew, scaleh] = video.get_dpi_scale_factor();
```

Comment:
The variable seems to be used for add-ons, e.g. we set a WML variable: https://github.com/wesnoth/wesnoth/blob/master/src/gui/widgets/helper.cpp#L104

We should probably also make sure that the settings can be "live-updated" without restarting Wesnoth. Meaning that the WML variable gets updated on Scale change. I'll test it out after building. (I'd also easily see it as something that could be finished up later if it's not trivial, as having HiDPI support in itself is already great.)

Reply:
It's not a WML variable, it's a WFL (formula language) variable. Considering if we should remove the `screen_pitch_microns` stuff now that we have this, since IIRC, it was an attempt to implement stopgap HDPI scaling 🤔

Reply:
I absolutely think it's great to have a scheme to improve the playability of wesnoth on hidpi screens, and so I am all for this scaling option, but this approach is no less a stopgap than anything that's come before -- the only "real" solution is to stop "thinking" in terms of pixels internally and do all display in perceptual-scale units (degrees in field of vision, for one example, look at what modern browsers define a "px" as). That way text can be placed appropriately and in readable sizes, while the high resolution can be used for detail in the scenes and icons etc, if/when renderings at high pixel densities are available. Simply scaling all pixels is a stopgap for the former (text readability) that cuts off the latter opportunity (actually using the higher resolution for detail where it would add to the game experience).

Reply:
Though it's also somewhat unlikely that we'd be able to overhaul/replace all the art necessary for the latter.

Reply:
Hmm, if those specs are being made available to addons it should probably be expanded to add a couple more. Like the stuff i added to the build info page - there technically now is a "drawing size" in pixels which is what wesnoth will see and draw to, and there's a "window size" in desktop/display/screen coordinates which is used for changing resolution, and there's an "output size" in pixels of the final rendering target which may be even bigger for apple-style high dpi. I changed this "screen size" here to now refer to the "drawing size", so anything that uses that should still work at least.

Re: what gwhitney says, i agree that the basic problem is that everything works in pixels and there are (at least) three different meanings to a pixel nowdays. However i did move things in the correct direction by changing to use render textures. It should be possible (with some more work) to render some parts of the display in high-dpi, and the others in low/game-native dpi. A good place to start would be the background image (you can see in the screenshots i posted it's much higher quality in the non-scaled pic).

The biggest barrier to this actually isn't the rendering system but the input and event handling everywhere which assumes that input coordinates = render coordinates = display coordinates = output coordinates. I tried to at least make this distinction clear now in the rendering system.

Reply:
Yeah, https://github.com/wesnoth/wesnoth/tree/old_master_20181015 was actually a previous attempt to get hardware acceleration working, and IIRC it also stalled out over the same general issue.

Reply:
Bringing it back to the code in question - i also don't actually know what this function returns for high-dpi displays. The actual dot-pitch of the physical display? Or the dot pitch in the scaled windowing system coordinates?

In my case it returns the physical DPI of the 55" display (~80dpi), which won't ever trigger high dpi handling even though the pixels appear very small from where i sit. Given that, i'd say the `screen_pitch_microns` variable should be removed in favour of checking the drawing resolution, and behaving differently if it's large. That will still work here.

> We should probably also make sure that the settings can be "live-updated" without restarting Wesnoth. Meaning that the WML variable gets updated on Scale change

I checked briefly and it looks like this should update automatically. But let me know if it doesn't and i'll try to fix it.

Reply:
I'm more or less okay with removing `screen_pitch_microns` as long as it's only used internally in GUI2 formulas.

Reply:
I'll update the comment to reflect that it should probably be deprecated. Physical DPI is never a good way of determining user-apparent pixel size, as displays do need to return accurate physical DPI info to correctly size physically-relevant things (like PDFs). I checked the calculations in `data/gui/*` and they should definitely be updated to use a different method, such as comparing to the maximum supported window size.

Reply:
For now i updated `get_dpi_scale_factor()` as well to include pixel scale. I think this will keep behaviour approximately what it was. As i mentioned it doesn't trigger for me anyway so it's a little difficult to test.